### ISSUE TITLE
Evonorm-s: incorrect variance used to calculate normalization?

### ISSUE THREAD

**Issue by gwern** [tensorflow/tpu#871]
re https://github.com/tensorflow/tpu/commit/95647b535084fe369bdc9c12bace726cb91970b4#diff-16057781e361556b2798753ef5e08b5c

Evonorm-s is a layer normalization replacement for Batchnorm found by evolutionary/random search; the goal is to replace Batchnorm, which tries to standardize each layers' outputs into smaller outputs following a N(0,1) distribution overall.

Batchnorm relies on hidden global state which causes difficulty in reasoning about model behavior and can lead to nasty bugs when the global state is improperly created or applied to a different model, etc. In BigGAN in particular, because of how the global state interacts with GAN dynamics, it has to be estimated from scratch each minibatch, making BN quality highly sensitive to batch size. We have been bitten by BN bugs and confusion, and would like to replace it with something less treacherous and batchsize-dependent, and were looking at Evonorm-s. Evonorm-s promises to get the same (or better) quality results, while relying only on sample statistics: mean, variance, etc, without learning any batch normalization stats (as I understand [the Evonorm-s paper](https://arxiv.org/abs/2004.02967 "'Evolving Normalization-Activation Layers', Liu et al 2020")).

Our first attempts back in June failed, leading to weird diverged BigGAN runs (for some reason, Evonorm-S tends to collapse into green images). This was odd because Evonorm-s appears quite simple and we checked against the implementation in tensorflow/tpu as well. We have been revisiting it now that our deadlines have been extended, and taking a second look. 

`tf.nn.moments` is the local variance:

~~~
def group_std(x, groups=32, eps=1e-5):
  N, H, W, C = x.shape
  x = tf.reshape(x, [N, H, W, groups, C // groups])
  _, var = tf.nn.moments(x, [1, 2, 4], keepdims=True)
  std = tf.sqrt(var + eps)
  std = tf.broadcast_to(std, x.shape)
  return tf.reshape(std, [N, H, W, C])
~~~

And the core of Evonorm-s is:

~~~
                          num = x * tf.nn.sigmoid(self.v * x)
                          return num / group_std(x) * self.gamma + self.beta
~~~

Which we might pseudo-code as

~~~
m = x * sigmoid(learned_constant * x)
m /= variance
turn num * gamma + beta
~~~

His question: what is the `/=` variance supposed to be?
On our TPU pods, this seems to not take the cross replica variance, but only the local (per-core) batch variance, which is like... two. (Instead of over _n_=1024.)

We switched the variance, and we seem to be getting better results after that.

So our question is: is the use of local per-core variance there correct and intentional? Has this specific implementation been tested on BigGAN or other large-scale problems and verified to work as well as the paper? (The paper is unclear but we *think*, based on mooch's comments elsewhere and the fact that it mentions GPUs rather than TPUs and the different FID/IS numbers for biggan-deep, that the Evonorm paper is using implementations either in PyTorch or JAX, and that this Tensorflow TPU implementation is a reimplementation and so was not necessarily tested by the paper authors.) The only test in that code is a pretty minor test of loading, not even training to convergence.

**Comment by rxsang** [comment]
By default, TPUs is doing things locally unless users explicitly reduce the results from all the replicas, which aligns with your observation which the variance is local. If you need the global variance, users need to handle the math correctly by explicitly all reducing over all TPU cores using `tf.compat.v1.tpu.cross_replica_sum` (if you are in TF1) or `strategy.reduce` (if you are in TF2). You can actually look at the implementation of `tf.keras.layers.experimental.SyncBatchNormalization` for reference https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/layers/normalization_v2.py#L168
